File: typescript-sdk/src/session.ts
import type { LedgerClient } from "./client.js";
import type { BatchAck, RecordAck, SessionSubmitInput } from "./types.js";
import { newRecordId } from "./utils.js";

// ---------------------------------------------------------------------------
// Session
//
// A lightweight wrapper that binds a session_id to a LedgerClient, so
// callers don't have to pass it on every record. Purely client-side sugar;
// no server-side session lifecycle exists.
// ---------------------------------------------------------------------------

export class Session {
  /** The bound session_id. Read-only. */
  readonly id: string;

  private readonly client: LedgerClient;

  constructor(client: LedgerClient, sessionId?: string) {
    this.client = client;
    this.id = sessionId ?? newRecordId();
  }

  /**
   * Submit a single record, auto-injecting `session_id = this.id`.
   * All other auto-fill rules (agent_id, record_id, schema_version,
   * client_ts_utc) apply as on LedgerClient.submit.
   */
  submit(input: SessionSubmitInput): Promise<RecordAck> {
    return this.client._submit({ ...input, session_id: this.id });
  }

  /**
   * Submit a batch of records, auto-injecting `session_id = this.id`
   * on each record. All other auto-fill rules apply per record.
   */
  submitBatch(inputs: SessionSubmitInput[]): Promise<BatchAck> {
    return this.client._submitBatch(inputs.map((input) => ({ ...input, session_id: this.id })));
  }
}

File: python-sdk/src/reasoning_ledger/session.py
from __future__ import annotations

from typing import TYPE_CHECKING, Any

from .types import BatchAck, RecordAck, SubmitInput
from .utils import new_record_id

if TYPE_CHECKING:
    from .client import LedgerClient


# ---------------------------------------------------------------------------
# Session
#
# A lightweight wrapper that binds a session_id to a LedgerClient, so
# callers don't have to pass it on every record. Purely client-side sugar;
# no server-side session lifecycle exists.
# ---------------------------------------------------------------------------


class Session:
    """A session bound to a specific session_id."""

    def __init__(self, client: LedgerClient, session_id: str | None = None) -> None:
        self._client = client
        self.id: str = session_id if session_id is not None else new_record_id()

    def submit(self, input_record: dict[str, Any]) -> RecordAck:
        """Submit a single record, auto-injecting session_id = self.id."""
        return self._client._submit({**input_record, "session_id": self.id})

    def submit_batch(self, inputs: list[SubmitInput]) -> BatchAck:
        """Submit a batch, auto-injecting session_id = self.id on each record."""
        return self._client._submit_batch(
            [{**inp, "session_id": self.id} for inp in inputs]
        )
